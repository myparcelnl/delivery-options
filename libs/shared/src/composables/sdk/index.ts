export * from './useCarrierRequest';
export * from './useCarriersRequest';
export * from './useDeliveryOptionsRequest';
export * from './usePickupLocationsRequest';
export * from './useRequest';
export * from './useRequestClient';
