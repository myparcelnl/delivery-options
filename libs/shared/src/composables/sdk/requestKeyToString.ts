import {isNumber, isString} from 'radash';
import {type RequestKey} from '../../types';

export const requestKeyToString = (queryKey: string | number | object | RequestKey): string => {
  if (isString(queryKey) || isNumber(queryKey)) {
    return String(queryKey);
  }

  if (Array.isArray(queryKey)) {
    return queryKey.map(requestKeyToString).join(':');
  }

  return JSON.stringify(queryKey);
};
