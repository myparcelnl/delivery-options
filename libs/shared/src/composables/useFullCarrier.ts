import {type MaybeRef, type Ref} from 'vue';
import {asyncComputed, get, useMemoize} from '@vueuse/core';
import {getFullCarrier} from '../utils';
import {type CarrierIdentifier, type FullCarrier, type SupportedPlatformName} from '../types';
import {resolveRefKey} from '../resolveRefKey';

const cb = (
  carrierIdentifier: MaybeRef<CarrierIdentifier>,
  platformName: MaybeRef<SupportedPlatformName>,
): Ref<FullCarrier> => {
  return asyncComputed(() => getFullCarrier(get(carrierIdentifier), get(platformName)));
};

export const useFullCarrier = useMemoize(cb, {getKey: resolveRefKey});
