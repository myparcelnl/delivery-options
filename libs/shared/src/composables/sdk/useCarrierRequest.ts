import {type MaybeRef} from 'vue';
import {get} from '@vueuse/core';
import {type EndpointResponse, type GetCarrier} from '@myparcel/sdk';
import {type CarrierName} from '@myparcel/constants';
import {type RequestHandler} from '../../types';
import {REQUEST_KEY_CARRIERS} from '../../data';
import {useSdk} from '..';
import {useRequest} from './useRequest';

export const useCarrierRequest = (
  carrier: MaybeRef<CarrierName>,
): RequestHandler<EndpointResponse<GetCarrier>[number]> => {
  return useRequest([REQUEST_KEY_CARRIERS, get(carrier)], async () => {
    const sdk = useSdk();

    return (await sdk.getCarrier({path: {carrier: get(carrier)}}))?.[0];
  });
};
