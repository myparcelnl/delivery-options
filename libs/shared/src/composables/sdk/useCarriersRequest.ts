import {type EndpointResponse, type GetCarriers} from '@myparcel/sdk';
import {type RequestHandler} from '../../types';
import {REQUEST_KEY_CARRIERS} from '../../data';
import {useSdk} from '..';
import {useRequestStorage} from './useRequestStorage';
import {useRequest} from './useRequest';

export const useCarriersRequest = (): RequestHandler<EndpointResponse<GetCarriers>> => {
  return useRequest(
    [REQUEST_KEY_CARRIERS],
    async () => {
      const sdk = useSdk();

      return sdk.getCarriers();
    },
    {
      onSuccess(carriers) {
        const requestStorage = useRequestStorage();

        /**
         * Update the cache for each carrier.
         */
        carriers.forEach((carrier) => {
          requestStorage.set([REQUEST_KEY_CARRIERS, carrier.name], carrier);
        });
      },
    },
  );
};
