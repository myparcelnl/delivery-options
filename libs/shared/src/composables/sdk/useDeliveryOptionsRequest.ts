import {type EndpointParameters, type EndpointResponse, type GetDeliveryOptions} from '@myparcel/sdk';
import {useSdk} from '../useSdk';
import {type RequestHandler} from '../../types';
import {REQUEST_KEY_DELIVERY_OPTIONS} from '../../data';
import {useRequest} from './useRequest';

export const useDeliveryOptionsRequest = (
  parameters: EndpointParameters<GetDeliveryOptions>,
): RequestHandler<EndpointResponse<GetDeliveryOptions>> => {
  return useRequest([REQUEST_KEY_DELIVERY_OPTIONS, parameters], async () => {
    const sdk = useSdk();

    try {
      return await sdk.getDeliveryOptions({parameters});
    } catch (e) {
      return [];
    }
  });
};
