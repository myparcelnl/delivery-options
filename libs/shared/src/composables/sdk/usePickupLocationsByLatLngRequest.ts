import {type EndpointParameters, type GetPickupLocations, type PickupLocation} from '@myparcel/sdk';
import {useSdk} from '../useSdk';
import {type RequestHandler} from '../../types';
import {REQUEST_KEY_PICKUP_LOCATIONS} from '../../data';
import {useRequest} from './useRequest';

export const usePickupLocationsByLatLngRequest = (
  parameters: EndpointParameters<GetPickupLocations>,
): RequestHandler<PickupLocation[]> => {
  return useRequest(
    [REQUEST_KEY_PICKUP_LOCATIONS, parameters],
    () => {
      const sdk = useSdk();

      return sdk.getPickupLocations({parameters});
    },
    {
      fallback: [],

      /**
       * TODO: Remove this when the api returns a proper response rather than 500 internal server error when no pickup
       *  locations are found (API-1301)
       */
      onError() {
        // silently fail
      },
    },
  );
};
