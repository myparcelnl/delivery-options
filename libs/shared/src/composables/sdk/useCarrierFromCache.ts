import {type MaybeRef, computed, toValue} from 'vue';
import {type CarrierName} from '@myparcel/constants';
import {type EndpointResponse, type GetCarrier} from '@myparcel/sdk';
import {useCarriersRequest} from './useCarriersRequest';
import {type RequestHandler} from '../../types';

export const useCarrierFromCache = (
  carrier: MaybeRef<CarrierName>,
): RequestHandler<EndpointResponse<GetCarrier>[number] | undefined> => {
  const carriersRequest = useCarriersRequest();
  
  const data = computed(() => {
    const carrierName = toValue(carrier);
    if (!carrierName || !carriersRequest.data.value) return undefined;
    
    return carriersRequest.data.value.find(c => c.name === carrierName);
  });

  const loading = computed(() => carriersRequest.loading.value);

  const load = async () => {
    await carriersRequest.load();
  };

  return {
    data,
    loading,
    load,
  };
};
