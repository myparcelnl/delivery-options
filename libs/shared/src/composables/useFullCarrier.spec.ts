/* eslint-disable max-nested-callbacks */
import {beforeEach, describe, expect, it} from 'vitest';
import {createPinia, setActivePinia} from 'pinia';
import {
  CarrierId,
  CarrierName,
  DeliveryTypeName,
  PackageTypeName,
  PlatformName,
  ShipmentOptionName,
} from '@myparcel/constants';
import {type CarrierIdentifier, type SupportedPlatformName} from '../types';
import {CarrierSetting} from '../data';
import {useFullCarrier} from './useFullCarrier';
import {useCarrierRequest} from './sdk';

async function loadCarriers(): Promise<void> {
  await useCarrierRequest(CarrierName.PostNl).load();
  await useCarrierRequest(CarrierName.DhlForYou).load();
}

describe('useFullCarrier', () => {
  beforeEach(() => {
    setActivePinia(createPinia());
  });

  it.each([
    [CarrierName.PostNl, PlatformName.MyParcel],
    [`${CarrierName.PostNl}:12345`, PlatformName.MyParcel],
    [CarrierName.PostNl, PlatformName.SendMyParcel],
    [`${CarrierName.PostNl}:12345`, PlatformName.SendMyParcel],
  ] satisfies [CarrierIdentifier, SupportedPlatformName][])(
    'returns carrier data for %s on platform %s',
    async (identifier, platform) => {
      expect.assertions(6);

      const postNl = useFullCarrier(identifier, platform);

      // Still loading
      expect(postNl.value).toBeUndefined();

      await loadCarriers();

      expect(postNl.value).toBeDefined();
      expect(postNl.value.id).toBe(CarrierId.PostNl);
      expect(postNl.value.name).toBe(CarrierName.PostNl);
      expect(postNl.value.human).toBeTypeOf('string');
      expect(postNl.value.meta).toBeInstanceOf(Object);
    },
  );

  describe('exposes utility methods', () => {
    it('can use hasPackageType', async () => {
      expect.assertions(2);

      const postNl = useFullCarrier(CarrierName.PostNl, PlatformName.MyParcel);
      const dhlForYou = useFullCarrier(CarrierName.DhlForYou, PlatformName.MyParcel);

      await loadCarriers();
      expect(postNl.value.hasPackageType(PackageTypeName.DigitalStamp)).toBe(true);
      expect(dhlForYou.value.hasPackageType(PackageTypeName.DigitalStamp)).toBe(false);
    });

    it('can use hasDeliveryType', async () => {
      expect.assertions(2);

      const postNl = useFullCarrier(CarrierName.PostNl, PlatformName.MyParcel);
      const dhlForYou = useFullCarrier(CarrierName.DhlForYou, PlatformName.MyParcel);

      await loadCarriers();
      expect(postNl.value.hasDeliveryType(DeliveryTypeName.Morning)).toBe(true);
      expect(dhlForYou.value.hasDeliveryType(DeliveryTypeName.Morning)).toBe(false);
    });

    it('can use hasShipmentOption', async () => {
      expect.assertions(2);

      const postNl = useFullCarrier(CarrierName.PostNl, PlatformName.MyParcel);
      const dhlForYou = useFullCarrier(CarrierName.DhlForYou, PlatformName.MyParcel);

      await loadCarriers();
      expect(postNl.value.hasShipmentOption(ShipmentOptionName.OnlyRecipient)).toBe(true);
      expect(dhlForYou.value.hasShipmentOption(ShipmentOptionName.OnlyRecipient)).toBe(true);
    });

    it('can use hasFeature', async () => {
      expect.assertions(2);

      const postNl = useFullCarrier(CarrierName.PostNl, PlatformName.MyParcel);
      const dhlForYou = useFullCarrier(CarrierName.DhlForYou, PlatformName.MyParcel);

      await loadCarriers();
      expect(postNl.value.hasFeature(CarrierSetting.AllowMondayDelivery)).toBe(true);
      expect(dhlForYou.value.hasFeature(CarrierSetting.AllowMondayDelivery)).toBe(false);
    });

    it('can use hasDeliveryInCountry', async () => {
      expect.assertions(2);

      const postNl = useFullCarrier(CarrierName.PostNl, PlatformName.MyParcel);
      const dhlForYou = useFullCarrier(CarrierName.DhlForYou, PlatformName.MyParcel);

      await loadCarriers();
      expect(postNl.value.hasDeliveryInCountry('NL')).toBe(true);
      expect(dhlForYou.value.hasDeliveryInCountry('NL')).toBe(true);
    });

    it('can use hasPickupInCountry', async () => {
      expect.assertions(2);

      const postNl = useFullCarrier(CarrierName.PostNl, PlatformName.MyParcel);
      const dhlForYou = useFullCarrier(CarrierName.DhlForYou, PlatformName.MyParcel);

      await loadCarriers();
      expect(postNl.value.hasPickupInCountry('BE')).toBe(true);
      expect(dhlForYou.value.hasPickupInCountry('BE')).toBe(false);
    });
  });
});
