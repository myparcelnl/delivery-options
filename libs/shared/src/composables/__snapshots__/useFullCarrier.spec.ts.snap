// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useFullCarrier > returns the carrier configuration for "bol.com" on "belgie" 1`] = `"❌"`;

exports[`useFullCarrier > returns the carrier configuration for "bol.com" on "belgie" 2`] = `"❌"`;

exports[`useFullCarrier > returns the carrier configuration for "bol.com" on "myparcel" 1`] = `"❌"`;

exports[`useFullCarrier > returns the carrier configuration for "bol.com" on "myparcel" 2`] = `"❌"`;

exports[`useFullCarrier > returns the carrier configuration for "bpost" on "belgie" 1`] = `"❌"`;

exports[`useFullCarrier > returns the carrier configuration for "bpost" on "myparcel" 1`] = `"❌"`;

exports[`useFullCarrier > returns the carrier configuration for "cheapcargo" on "belgie" 1`] = `"❌"`;

exports[`useFullCarrier > returns the carrier configuration for "cheapcargo" on "myparcel" 1`] = `"❌"`;

exports[`useFullCarrier > returns the carrier configuration for "dhl" on "belgie" 1`] = `"❌"`;

exports[`useFullCarrier > returns the carrier configuration for "dhl" on "myparcel" 1`] = `"❌"`;

exports[`useFullCarrier > returns the carrier configuration for "dhleuroplus" on "belgie" 1`] = `"❌"`;

exports[`useFullCarrier > returns the carrier configuration for "dhleuroplus" on "myparcel" 1`] = `"❌"`;

exports[`useFullCarrier > returns the carrier configuration for "dhlforyou" on "belgie" 1`] = `"❌"`;

exports[`useFullCarrier > returns the carrier configuration for "dhlforyou" on "myparcel" 1`] = `"❌"`;

exports[`useFullCarrier > returns the carrier configuration for "dhlparcelconnect" on "belgie" 1`] = `"❌"`;

exports[`useFullCarrier > returns the carrier configuration for "dhlparcelconnect" on "myparcel" 1`] = `"❌"`;

exports[`useFullCarrier > returns the carrier configuration for "dpd" on "belgie" 1`] = `"❌"`;

exports[`useFullCarrier > returns the carrier configuration for "dpd" on "myparcel" 1`] = `"❌"`;

exports[`useFullCarrier > returns the carrier configuration for "instabox" on "belgie" 1`] = `"❌"`;

exports[`useFullCarrier > returns the carrier configuration for "instabox" on "myparcel" 1`] = `"❌"`;

exports[`useFullCarrier > returns the carrier configuration for "postnl" on "belgie" 1`] = `"❌"`;

exports[`useFullCarrier > returns the carrier configuration for "postnl" on "myparcel" 1`] = `"❌"`;

exports[`useFullCarrier > returns the carrier configuration for "ups" on "belgie" 1`] = `"❌"`;

exports[`useFullCarrier > returns the carrier configuration for "ups" on "myparcel" 1`] = `"❌"`;
